File: java/CircularLinkedList.java
class Node {
  int data;
  Node next;

  public Node(int data) {
    this.data = data;
    this.next = null;
  }
}

public class CircularLinkedList {
  Node head;

  public CircularLinkedList() {
    head = null;
  }

  // insert node at beginning
  public void insertAtBeginning(int data) {
    Node newNode = new Node(data);
    if(head == null) {
      head = newNode;
      newNode.next = head;
    } else {
      Node last = head;
      while(last.next != head) {
        last = last.next;
      }
      newNode.next = head;
      last.next = newNode;
      head = newNode;
    }
  }

  // insert node at end
  public void insertAtEnd(int data) {
    Node newNode = new Node(data);
    if(head == null) {
      head = newNode;
      newNode.next = head;
    } else {
      Node last = head;
      while(last.next != head) {
        last = last.next;
      }
      last.next = newNode;
      newNode.next = head;
    }
  }

  // delete node by key
  public void deleteByKey(int key) {
    if(head == null) {
      return;
    }
    Node prev = null;
    Node current = head;
    do {
      if(current.data == key) {
        if(prev == null) {
          if(head.next == head) {
            head = null;
          } else {
            head = current.next;
            prev = head;
            while(prev.next != current) {
              prev = prev.next;
            }
            prev.next = head;
          }
          return;
        } else {
          prev.next = current.next;
          return;
        }
      }
      prev = current;
      current = current.next;
    } while (current != head);
  }

  // print list
  public void printList() {
    if(head == null) {
      return;
    }
    Node current = head;
    do {
      System.out.println(current.data + " ");
      current = current.next;
    } while (current != head);
    System.out.println();
  }
}

File: javascript/CircularLinkedList.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class CircularLinkedList {
  constructor() {
    this.head = null;
  }
  isEmpty() {
    return this.head === null;
  }
  add(data) {
    const newNode = new Node(data);
    if(this.isEmpty()) {
      this.head = newNode;
      newNode.next = newNode;
    } else {
      let current = this.head;
      while(current.next !=== this.head) {
        current = current.next;
      }
      current.next = newNode;
      newNode.next = this.head;
    }
  }

  delete(data) {
    if(this.isEmpty()) {
      return;
    }
    let current = this.head, prev = null;
    if(current.data === data) {
      if(current.next === current) {
        this.head = null;
      } else {
        while(current.next !== this.head) {
          current = current.next;
        }
        current.next = this.head.next;
        this.head = this.head.next;
      }
      return;
    }
    while(current.next !== this.head && current.next.data !== data) {
      current = current.next;
    }
    if(current.next.data === data) {
      current.next = current.next.next;
    }
  }

  printList() {
    if(this.isEmpty()) {
      console.log("list is empty");
      return;
    }
    let current = this.head;
    do {
      console.log(current.data);
      current = current.next;
    } while (current !== this.head);
  }
}
